btrfs-progs: mkfs: add --compress

maharmstone opened the pull request:
Adds a --compress option to mkfs.btrfs when using --rootdir. (Resurrection of https://github.com/kdave/btrfs-progs/pull/882, closed by mistake.)

@adam900710, this includes your two patches, and all the suggestions that you made. If you're happy with all this, I suggested accepting this PR and abandoning https://github.com/kdave/btrfs-progs/pull/919.

adam900710 commented:
Thanks a lot. I'll close #882 and use this branch instead. 

Although to avoid previous accidents l'll take a more comprehensive review before merge.
So far the patches looks good to me (through the github WebUI) and I believe we can push this for the next cycle.

adam900710 commented on convert/source-reiserfs.c:
```diff
@@ -384,7 +384,8 @@ static int reiserfs_convert_tail(struct btrfs_trans_handle *trans,
 	ret = btrfs_insert_inline_extent(trans, root, objectid,
-					 offset, body, length);
+					 offset, body, length,
+					 0, length);
```
Please use `BTRFS_COMPRESS_NONE` to replace immediate number `0`.

adam900710 reviewed (commented):
Most small style problems, but the compression parameter and the inode flag handling is not subtle, and affecting all the other compression algorithms.

maharmstone commented on mkfs/rootdir.c:
```diff
@@ -455,6 +459,65 @@ static int insert_reserved_file_extent(struct btrfs_trans_handle *trans,
+	 * Try to compress the first sector - if it would be larger,
+	 * return -E2BIG.
+	 */
+	if (!(btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_COMPRESS)) {
```
>Furthermore, if we got one 1MiB extent that its first sector is not compressible but overall it can still be compressed, then I'd say we lose the chance to save some space.

The downside to this approach is that if a file is incompressible, you end up giving it extents of 128KB rather than 1MB. Files that are already compressed are going to be more common than the worst-case scenario of a low-entropy file with a high-entropy first sector.

maharmstone commented on mkfs/rootdir.c:
```diff
@@ -455,6 +459,65 @@ static int insert_reserved_file_extent(struct btrfs_trans_handle *trans,
+	if (!(btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_COMPRESS))
+		return -E2BIG;
+
+	return 0;
```
If deflate returns Z_OK when called with Z_FINISH, that means that there wasn't enough space left in the output buffer. This isn't a very intuitive part of the zlib spec.

The logic is sound here, but I'll make it more obvious that this isn't an oversight.

maharmstone commented:
Qu, can you please give this latest version a go?

In addition to the issues mentioned above, I've changed zlib_compress_inline_extent etc. so that like zlib_compress_extent etc. they return the size of the compressed data, rather than putting it in a pointer. I've also removed the confusing 0 / -E2BIG distinction, and put the logic in the caller instead.

adam900710 commented:
Looks very good to me now.

I'll merge it after manually verifying the resulted image. You may want to add another series just for the extra test cases.

adam900710 commented:
I manually checked the following points, and hopes it will help your next test cases:

- Super incompat flags for ZSTD and LZO (zlib is supported by default)
- Incompressible file detection (using `/dev/random`)
- Resulted file extents (proper zstd/lzo/zlib compression)
- Inode flags (should have compression flag)
- With ZSTD/LZO disabled, compile should pass and using that algo should error out
  LZO failed to compile, fixed in the `devel` branch now.
  Not sure how we can test the disabled algo, IIRC we have no previous example to test such disable features.
  (maybe NOTRUN if the feature is enabled?)

maharmstone commented:
Thanks Qu
